### PR TITLE
drivers: entropy: stm32: fix RNG clock source with all MSI frequencies

### PR DESCRIPTION
**Pull request by giansta**
Take account of all possible configured MSI frequencies when configuring
PLLSAI1 with MSI input to get the 48 MHz output clock.
If MSI input frequency is too low to obtain 48 MHz output, the highest
output frequency is chosen to be used by RNG: such solution should be
valid when USB is not used.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

### CODE REVIEW

File: drivers/entropy/entropy_stm32.c [hunk 1]
```diff
@@ -357,6 +357,18 @@ static int entropy_stm32_rng_init(const struct device *dev)
...
+#if !defined(CONFIG_CLOCK_STM32_PLL_SRC_MSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
+	/* If no PLLSRC clock, use other available 48 MHz clock source */
+#if defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI) && \
+	defined(CONFIG_CLOCK_STM32_MSI_RANGE) && \
+	CONFIG_CLOCK_STM32_MSI_RANGE == 11
+	LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_MSI);
+#else
+	LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_HSI48);
```
Comment (ABOSTM) [inline drivers/entropy/entropy_stm32.c:369]: Not all STM32L4 have HSI48 as RNG clock source. For example reference man RM0351 for stm32L4x6.

File: drivers/entropy/entropy_stm32.c [hunk 2]
```diff
@@ -377,6 +389,7 @@ static int entropy_stm32_rng_init(const struct device *dev)
 	 *  Linear Feedback Shift Register
 	 */
 	 LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_PLLSAI1);
+#endif
```
Comment (ABOSTM) [inline drivers/entropy/entropy_stm32.c:392]: For an easy match with #if: 
```suggestion
#endif // CONFIG_CLOCK_STM32_PLL_SRC_MSI && !CONFIG_CLOCK_STM32_PLL_SRC_HSI) \
&& !(CONFIG_CLOCK_STM32_PLL_SRC_HSE
```

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Hi @giansta, thanks for your pull request

File: drivers/entropy/entropy_stm32.c (hunk 1 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Thank you for pointing this out.
In the rebased branch the HSI configuration (that is not expected to happen for STM32L47x/L48x devices) is checked first. In case no 48 MHz source clock is available an error is returned.

Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Even the #endifs should be cleaner now.

File: drivers/entropy/entropy_stm32.c [hunk 3]
```diff
@@ -357,6 +357,24 @@ static int entropy_stm32_rng_init(const struct device *dev)
...
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
+
+	/* If no PLLSRC clock, use other available 48 MHz clock source */
+#if defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)
+	LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_HSI48);
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI) && \
+	defined(CONFIG_CLOCK_STM32_MSI_RANGE) && \
+	CONFIG_CLOCK_STM32_MSI_RANGE == 11
+	LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_MSI);
```
Comment (ABOSTM) [inline drivers/entropy/entropy_stm32.c:371]: I dig a bit more and I have several remarks:
1) I could not find a restriction in reference manual of STM32L4, to have MSI clock at 48MHZ.
Any MSI value seems valid for RNG.
The only restriction is when USB is also used in parallel, as they share the same clock and USB require 48MHz.
But it is up to USB to manage this restriction. 
2) I am not sure to understand what was wrong with the previous code:
It seems there is a full configuration of PLLSAI1 to feed the RNG.
Maybe just miss to enable MSI (LL_RCC_MSI_Enable) ?
```
	/* Configure PLLSA11 to enable 48M domain */
	LL_RCC_PLLSAI1_ConfigDomain_48M(LL_RCC_PLLSOURCE_MSI,
					LL_RCC_PLLM_DIV_1,
					24, LL_RCC_PLLSAI1Q_DIV_2);

	/* Enable PLLSA1 */
	LL_RCC_PLLSAI1_Enable();

	/*  Enable PLLSAI1 output mapped on 48MHz domain clock */
	LL_RCC_PLLSAI1_EnableDomain_48M();

	/* Wait for PLLSA1 ready flag */
	while (LL_RCC_PLLSAI1_IsReady() != 1) {
	}

	/*  Write the peripherals independent clock configuration register :
	 *  choose PLLSAI1 source as the 48 MHz clock is needed for the RNG
	 *  Linear Feedback Shift Register
	 */
	 LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_PLLSAI1);
```

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: drivers/entropy/entropy_stm32.c [hunk 4]
```diff
@@ -357,6 +357,24 @@ static int entropy_stm32_rng_init(const struct device *dev)
...
+#else
+	/* No 48 MHz source clock available for RNG */
+	return -EINVAL;
```
[the remark this answers is not in the document]
Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Indeed this was my first choice. Then I saw another error case in the function and I changed, but it was a bad choice: this is not a runtime error.
Thanks.

File: drivers/entropy/entropy_stm32.c (hunk 3 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: 1. 1. Thanks for the input, the code is even simpler without MSI frequency restrictions.

Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: 2. In our STM32L462 system we chose to use MSI @ 48 MHz as system clock and we have both PLL and PLLSAI1 disabled.
This means having PLLSRC=00 ("No clock sent to PLL PLLSAI1" and even PLL) that corresponds having no CONFIG_CLOCK_STM32_PLL_SRC_xxx defined in Zephyr.
With the previous code, after programming PLLSAI1, the loop to wait for PLLSA1 ready flag was hanging.
Yesterday I was supposing that the previous code was not working just because we have no PLL programmed (that's why of compile switch "#if !defined(CONFIG_CLOCK_STM32_PLL_SRC_MSI) && ..."). 
After reading your note, I remembered that in the past days I already noticed that apparently the previous code should work, but we didn't find the reason why it's looping.
On the other hand, we don't want to enable PLL and we already have a 48 MHz source, so the natural decision was to use MSI.
Now I believe to have found why the previous code doesn't work: it probably assumes to have MSI @ 4 MHz and it's multiplying it by 24 to have 48 MHz; instead it's multiplying 48 MHz *24!
In conclusion, the current solution uses only the existing configuration to obtain the following logic: if no PLL is wanted, either use MSI or HSI as clock source for RNG, otherwise use PLL.

File: drivers/entropy/entropy_stm32.c [hunk 5]
```diff
@@ -357,6 +357,21 @@ static int entropy_stm32_rng_init(const struct device *dev)
 	__ASSERT_NO_MSG(dev_cfg != NULL);
 
 #if CONFIG_SOC_SERIES_STM32L4X
+
+#if !defined(CONFIG_CLOCK_STM32_PLL_SRC_MSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) && \
+	!defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
```
Comment (erwango) [inline drivers/entropy/entropy_stm32.c:363]: This is equivalent to `!CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL`, which is exclusive with CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI and CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI.
So you don't need these 2 levels of #if's

File: drivers/entropy/entropy_stm32.c [hunk 6]
```diff
@@ -357,6 +357,21 @@ static int entropy_stm32_rng_init(const struct device *dev)
...
+
+	/* If no PLLSRC clock, use other available 48 MHz clock source */
+#if defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)
+	LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_HSI48);
```
Comment (erwango) [inline drivers/entropy/entropy_stm32.c:367]: HSI and HSI48 are 2 different clocks.
`CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI=y` doesn't mean HSI48 is on.
HSI48 requires dedicated activation.
Besides HSI48 is not available on all L4, so you should additionally check RCC_HSI48_SUPPORT.

This being said, would it be required ?
Since these are 2 different clocks, this is basically the same case as initially enabling a different clock (PLLSAI).

All in all, maybe you just need to treat the MSI case (since indeed only one clock is required) and the alternate case where an additional clock is required (PLLSAI). And fix what is missing to get it work (enable MSI if missing).

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/entropy/entropy_stm32.c (hunk 5 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Thanks for this input, it should simplify a lot.

File: drivers/entropy/entropy_stm32.c (hunk 6 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Thanks for your corrections and suggestions. If I understand correctly, you're suggesting either to use MSI, without any PLL, or to us PLLSAI1. Is this correct? I'm drafting this in latest commit.
Anyway, I would add some considerations and possibly additional fixes to find a common solution.

- I was trying to reuse existing configs rather than introducing new ones, but I was probably wrong: the CLK48SEL bits select the RNG/USB clock that is well different than SYSCLK selection, so we would need a new Kconfig here, do you agree?
- In such case anyway the logic and the code would become more complex and depend also on SYSCLK selection. E.g. selecting MSI for RNG source would imply different code if SYSCLK is MSI or not.
- Perhaps a simpler logic like the one we're using now would be preferable?
- The PLLSAI1 usage in any case needs to be fixed with valid multiplier/divisors for different cases, e.g. if PLLSRC is MSI @ 48 MHz is not working now.

Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Besides the above points, in the past days we found other problems. 
My colleagues found that the solution with MSI only @ 48 MHz (working both as SYSCLK and  CLK48SEL for RNG) was generating continuously interrupts. As soon as RNG was enabled, it started generating interrupts with CEIS=1, preventing the application to run. This was happening sometimes, the conditions are not clarified (depending on HW?), but the initialization is always the same.
Reading the Reference manual (e.g section 24.3.7), our current code and drivers code, we found some points that are still not clear.
- The CED (bit 5 in RNG CR) seems not available in STM32L462, so we're assuming that the condition CED = 0 is always true for our platform.
- The RNG clock frequency should never be too low, as we first initialize the MSI in stm32_clock_control_init() and then we use it in entropy_stm32_rng_init(). The condition "rng_clk > rng_hclk/16" should always be true (48>48/16). Some delay in between would help?
- Even if we don't know why we get the interrupt with this condition, the stm32_rng_isr() seemed not to clear the CEIS bit interrupt flag like the RM recommends. In my last commit I tried to fix the error case handling, we're testing it from now on.
- Probably lower prio fix to do: the random_byte_get() function appears is returning valid read RNG values that are interpreted as errors because negative.
- It's unclear to us the statement in table 115 that rng_clk should be asynchronous to rng_hclk: is this a real requirement? Can this be a cause for this error?
- Because of the above asynchronous requirement, we tried using PLLSAI1 with MSI @ 48 MHZ as source, using different multiplicator and divisor to get 48 MHz output for RNG. It seemed to work for some days (even if it appears to be still synchronous), but today we found that the same interrupt with CEIS=1 error still happens.
- Our SYSCLK configuration with MSI is using the MSIPLLEN (auto calibration with LSE, PLL-mode). This helped a lot in having a precise clock, hoping not generating issues like this one.

Comment (erwango) [reply drivers/entropy/entropy_stm32.c]: @giansta, would you mind opening a dedicated PR/Issue for this specific change ?
I know both topics are linked, but from what I understand they could be dealt with separately.
This will help merging faster in the end

Comment (giansta) [reply drivers/entropy/entropy_stm32.c]: Do you mean a dedicated PR/Issue for the flags clearing in the ISR?

Comment (erwango) [reply drivers/entropy/entropy_stm32.c]: Yes

**Comment by erwango** [comment]
@giansta, do you plan to continue this PR?

**Comment by giansta** [comment]
> 
> 
> @giansta, do you plan to continue this PR?

Yes, it's in my queue. In next days I'll restart with it, splitting a couple of minor changes in other PRs, as you proposed, and proposing a solution.

**Comment by giansta** [comment]
Changes have been rebased starting from original issue: the PLLSAI1 configuration was working for the default case of MSI @ 4 Mhz, while it was looping forever waiting for PLLSA1 ready flag in case of MSI @ 48 MHz, because output PLL48M2CLK was programmed @ 576 MHz.

A couple of minor changes are going to be proposed in dedicated PRs, as suggested.

Besides that, another change would be needed adding a new Kconfig entry (e.g. "48 MHz clock source selection" in "Device Drivers - Hardware Clock Controller Support - STM32 Reset & Clock Control") to enable changing the source clocl for RNG & USB.
I would like to proceed with this in this PR, extending `entropy_stm32_rng_init_l4()`, yet I would like to know your point of view.

Finally, we have a workaround for the issue we had with generation of continuous interrupts that I mentioned above. We would like to discuss it, maybe the best way is to open an issue, isn't it?

**Comment by erwango** [comment]
> Finally, we have a workaround for the issue we had with generation of continuous interrupts that I mentioned above. We would like to discuss it, maybe the best way is to open an issue, isn't it?

Sure, it's easier to split unrelated topics to keep track of what is fixed and what remains to be.

**Review by FRASTM — APPROVED** [review]

**Comment by giansta** [comment]
A small change has been proposed in dedicated PR #3337.

When testing another small change for which I was thinking to create another PR, I discovered that with my new hardware and a more mature application that I could systematically reproduce the issue that my colleagues found time ago only on some boards in non clear circumstances.
Basically, when exiting from Stop 2 mode, we had continuous CEIS interrupts.
My "small change" wanted to clear interrupt flags in CEIS and SEIS cases, but I was neglecting the RM unclear statement "the application should check that the RNG clock is configured correctly". On the other end, workaround was un-configuring/re-configuring the entropy clocks when the event queue was full was full/empty below threshold.
Now I could verify that the workaround can be avoided if the clocks are re-configured all the times we find CEIS=1. The two function calls in this case are entropy_stm32_rng_init_l4() that re-configures the clocks and resets CECS and LL_RNG_ClearFlag_CEIS() that resets CEIS only. Without both calls our system couldn't run and watchdog fired.

Unknown facts:
 1. When removing entropy_stm32_rng_init_l4() call: in this case CEIS is reset and CECS is set, I don't see RNG interrupts but the system is blocked in some way and watchdog fires.
 2. Not sure about the exact cause that triggers CECS=1 when exiting Stop 2 mode.

In any case, our system cannot work only with the 1st commit of this PR, the second is needed as well.

Finally, these changes are L4-specific. I don't know exactly how to manage the other families and I cannot test them.

Another change still to make, mentioned above: new Kconfig entry to change the source clock for RNG & USB.

**Comment by giansta** [comment]
I believe that this PR is ready to be reviewed, isn't it?

File: drivers/entropy/entropy_stm32.c [hunk 7]
```diff
@@ -347,25 +347,55 @@ static int entropy_stm32_rng_get_entropy_isr(const struct device *dev,
...
+		}
+		msi_range >>= RCC_CSR_MSISRANGE_Pos;
+	} else {
+		msi_range = LL_RCC_MSI_GetRange();
+		if (msi_range > LL_RCC_MSIRANGE_11) {
+			return -EINVAL;
+		}
+		msi_range >>= RCC_CR_MSIRANGE_Pos;
+	}
+	__ASSERT_NO_MSG(msi_range < ARRAY_SIZE(pll_msi_cfg));
```
Comment (erwango) [inline drivers/entropy/entropy_stm32.c:417]: PLL should not be configured in this driver, this should be done in the dedicated driver (clock-control).
Also, for now for all PLLs we rely on user to provide the correct configuration using device tree, this should be the same for this one.

File: drivers/entropy/entropy_stm32.c [hunk 8]
```diff
@@ -381,7 +411,31 @@ static int entropy_stm32_rng_init(const struct device *dev)
...
+	__ASSERT_NO_MSG(dev != NULL);
+
+	dev_data = DEV_DATA(dev);
+	dev_cfg = DEV_CFG(dev);
+
+	__ASSERT_NO_MSG(dev_data != NULL);
+	__ASSERT_NO_MSG(dev_cfg != NULL);
+
+#if CONFIG_SOC_SERIES_STM32L4X
+	res = entropy_stm32_rng_init_l4();
```
Comment (erwango) [inline drivers/entropy/entropy_stm32.c:464]: Why is a L4 specific init required?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/entropy/entropy_stm32.c (hunk 7 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c:417]: The PLL configuration was already in this driver ("Configure PLLSA11 to enable 48M domain"). I just ha to fix it when using other MSI frequencies.
I'm not familiar with clock-control functions. Do you mean that the clock selection should be done in following clock_control_on() or something similar? Where is the PLLSAI1 clock selected in the device tree?
Should we solve this issue in this PR or does it make sense to move it out?

File: drivers/entropy/entropy_stm32.c (hunk 8 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c:464]: Not my decision, I've just extended it. There was already a L4 specific code in entropy_stm32_rng_init().

File: drivers/entropy/entropy_stm32.c (hunk 7 above)
Comment (erwango) [reply drivers/entropy/entropy_stm32.c:417]: > The PLL configuration was already in this driver ("Configure PLLSA11 to enable 48M domain"). I just ha to fix it when using other MSI frequencies.

Indeed, but this was just as a convenience. I'd like we put too much code for PLL support in drivers other than clocks.
This need to be centralized, for architecture coherency, ease of maintenance and footprint.

> I'm not familiar with clock-control functions. Do you mean that the clock selection should be done in following clock_control_on() or something similar? Where is the PLLSAI1 clock selected in the device tree?

We surely miss code for that now, but this should be added. Clock control description has been moved to dts recently as a preliminary step in that direction. I'd like to avoid having drivers going in the opposite direction from now on.

> Should we solve this issue in this PR or does it make sense to move it out?

This should probably be done in another PR

File: drivers/entropy/entropy_stm32.c (hunk 8 above)
Comment (erwango) [reply drivers/entropy/entropy_stm32.c:464]: I know this part of code would require some update, but I prefer to see evolution in the sense of factorization and commonalisation rather than splitting out per series.

File: drivers/entropy/entropy_stm32.c (hunk 7 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c:417]: I totally agree to have cleaner architecture, moving the code in proper drivers, have common code for different families, etc. We seem even to agree that the scope of this PR is fixing some non-working cases, limited to our experience with L462VE.

File: drivers/entropy/entropy_stm32.c (hunk 8 above)
Comment (giansta) [reply drivers/entropy/entropy_stm32.c:464]: Same observation as above.

**Comment by cfriedt** [comment]
@giansta - is this pr moving forward still?

**Comment by giansta** [comment]
> 
> 
> @giansta - is this pr moving forward still?

Yes, I'm working on it. I'm fixing my app to work with latest main branch and then try to fix RNG driver using the clock control tree, as suggested. I need few days, having an extended weekend.

**Comment by cfriedt** [comment]
@giansta - is this still moving forward?

**Comment by giansta** [comment]
> 
> 
> @giansta - is this still moving forward?

In the last two weeks I hoped to work on this, but I was too busy. In following two weeks I'm out, I foresee to be back on this on September 2nd half.

**Comment by cfriedt** [comment]
@giansta - are you able to complete this soon?
